<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Use Cases | Cosigt documentation</title>
    <meta name="generator" content="VuePress 1.9.10">
    
    <meta name="description" content="">
    
    <link rel="preload" href="/cosigtdoc/assets/css/0.styles.3e9b6cb4.css" as="style"><link rel="preload" href="/cosigtdoc/assets/js/app.ea2c52ed.js" as="script"><link rel="preload" href="/cosigtdoc/assets/js/3.d0293945.js" as="script"><link rel="preload" href="/cosigtdoc/assets/js/2.bc6597f8.js" as="script"><link rel="preload" href="/cosigtdoc/assets/js/1.50924563.js" as="script"><link rel="preload" href="/cosigtdoc/assets/js/14.3e4a7b1f.js" as="script"><link rel="preload" href="/cosigtdoc/assets/js/29.296314f5.js" as="script"><link rel="prefetch" href="/cosigtdoc/assets/js/10.cf03d8e5.js"><link rel="prefetch" href="/cosigtdoc/assets/js/11.8e68a12f.js"><link rel="prefetch" href="/cosigtdoc/assets/js/12.50592b84.js"><link rel="prefetch" href="/cosigtdoc/assets/js/13.374a5d22.js"><link rel="prefetch" href="/cosigtdoc/assets/js/15.da27eca8.js"><link rel="prefetch" href="/cosigtdoc/assets/js/18.fbf56f97.js"><link rel="prefetch" href="/cosigtdoc/assets/js/19.056f4829.js"><link rel="prefetch" href="/cosigtdoc/assets/js/20.65bea4bc.js"><link rel="prefetch" href="/cosigtdoc/assets/js/21.58f7b93d.js"><link rel="prefetch" href="/cosigtdoc/assets/js/22.71d7b881.js"><link rel="prefetch" href="/cosigtdoc/assets/js/23.91ce9f40.js"><link rel="prefetch" href="/cosigtdoc/assets/js/24.3a6270a5.js"><link rel="prefetch" href="/cosigtdoc/assets/js/25.3f12cfe1.js"><link rel="prefetch" href="/cosigtdoc/assets/js/26.8cb8c3ad.js"><link rel="prefetch" href="/cosigtdoc/assets/js/27.5dc334b8.js"><link rel="prefetch" href="/cosigtdoc/assets/js/28.3dd02697.js"><link rel="prefetch" href="/cosigtdoc/assets/js/30.41b87e58.js"><link rel="prefetch" href="/cosigtdoc/assets/js/31.e72a3f75.js"><link rel="prefetch" href="/cosigtdoc/assets/js/32.42c91560.js"><link rel="prefetch" href="/cosigtdoc/assets/js/33.ac668601.js"><link rel="prefetch" href="/cosigtdoc/assets/js/34.0463d7d5.js"><link rel="prefetch" href="/cosigtdoc/assets/js/35.24faa46f.js"><link rel="prefetch" href="/cosigtdoc/assets/js/36.f7952011.js"><link rel="prefetch" href="/cosigtdoc/assets/js/37.3936cd98.js"><link rel="prefetch" href="/cosigtdoc/assets/js/38.39491dd7.js"><link rel="prefetch" href="/cosigtdoc/assets/js/39.fcf0b573.js"><link rel="prefetch" href="/cosigtdoc/assets/js/4.d4f0fa8b.js"><link rel="prefetch" href="/cosigtdoc/assets/js/40.75445b96.js"><link rel="prefetch" href="/cosigtdoc/assets/js/41.4a66e0c6.js"><link rel="prefetch" href="/cosigtdoc/assets/js/42.36ab31b3.js"><link rel="prefetch" href="/cosigtdoc/assets/js/43.b3673f5a.js"><link rel="prefetch" href="/cosigtdoc/assets/js/44.5de79b32.js"><link rel="prefetch" href="/cosigtdoc/assets/js/45.72b0a3d1.js"><link rel="prefetch" href="/cosigtdoc/assets/js/46.66c1744e.js"><link rel="prefetch" href="/cosigtdoc/assets/js/47.8c8ded00.js"><link rel="prefetch" href="/cosigtdoc/assets/js/48.6d3f1d90.js"><link rel="prefetch" href="/cosigtdoc/assets/js/49.d6f16c9e.js"><link rel="prefetch" href="/cosigtdoc/assets/js/5.80b2fab8.js"><link rel="prefetch" href="/cosigtdoc/assets/js/50.6d501467.js"><link rel="prefetch" href="/cosigtdoc/assets/js/51.928baf40.js"><link rel="prefetch" href="/cosigtdoc/assets/js/52.26d55b7a.js"><link rel="prefetch" href="/cosigtdoc/assets/js/53.f71c9efb.js"><link rel="prefetch" href="/cosigtdoc/assets/js/54.8a3f53e6.js"><link rel="prefetch" href="/cosigtdoc/assets/js/55.37f9a0c0.js"><link rel="prefetch" href="/cosigtdoc/assets/js/6.7b0da6a8.js"><link rel="prefetch" href="/cosigtdoc/assets/js/7.d7421304.js"><link rel="prefetch" href="/cosigtdoc/assets/js/8.d8702868.js"><link rel="prefetch" href="/cosigtdoc/assets/js/9.3f6bcb54.js"><link rel="prefetch" href="/cosigtdoc/assets/js/vendors~docsearch.d23114a2.js">
    <link rel="stylesheet" href="/cosigtdoc/assets/css/0.styles.3e9b6cb4.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><!----> <header class="navbar"><div class="navbar-container"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/cosigtdoc/" class="home-link router-link-active"><!----> <span class="site-name">Cosigt documentation</span></a> <div class="links"><div class="search-box"><input aria-label="Search" placeholder="Search" autocomplete="off" spellcheck="false" value=""> <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class="search-icon"><path d="M21.7,20.3L18,16.6c1.2-1.5,2-3.5,2-5.6c0-5-4-9-9-9c-5,0-9,4-9,9c0,5,4,9,9,9c2.1,0,4.1-0.7,5.6-2l3.7,3.7c0.2,0.2,0.5,0.3,0.7,0.3s0.5-0.1,0.7-0.3C22.1,21.3,22.1,20.7,21.7,20.3zM4,11c0-3.9,3.1-7,7-7c3.9,0,7,3.1,7,7c0,1.9-0.8,3.7-2,4.9c0,0,0,0,0,0s0,0,0,0c-1.3,1.3-3,2-4.9,2C7.1,18,4,14.9,4,11z"></path></svg> <span class="search-command-wrapper search-command"><span class="search-command-char">⌘</span> <span class="search-command-char">K</span></span> <!----></div> <nav class="nav-links can-hide"><div class="nav-item"><a href="/cosigtdoc/" class="vp-link nav-link">
  Home
  </a></div><div class="nav-item"><a href="/cosigtdoc/contact/contact.html" class="vp-link nav-link">
  Contact
  </a></div><div class="nav-item"><a href="https://github.com/davidebolo1993/cosigt" target="_blank" class="vp-link external nav-link">
  GitHub
  <!----> </a></div> <div class="VPNavBarAppearance appearance" data-v-19d27b0e><button type="button" aria-label="toggle dark mode" class="vt-switch vt-switch-appearance" data-v-19d27b0e><span class="vt-switch-check"><span class="vt-switch-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class="vt-switch-appearance-sun"><path d="M12,18c-3.3,0-6-2.7-6-6s2.7-6,6-6s6,2.7,6,6S15.3,18,12,18zM12,8c-2.2,0-4,1.8-4,4c0,2.2,1.8,4,4,4c2.2,0,4-1.8,4-4C16,9.8,14.2,8,12,8z"></path> <path d="M12,4c-0.6,0-1-0.4-1-1V1c0-0.6,0.4-1,1-1s1,0.4,1,1v2C13,3.6,12.6,4,12,4z"></path> <path d="M12,24c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1s1,0.4,1,1v2C13,23.6,12.6,24,12,24z"></path> <path d="M5.6,6.6c-0.3,0-0.5-0.1-0.7-0.3L3.5,4.9c-0.4-0.4-0.4-1,0-1.4s1-0.4,1.4,0l1.4,1.4c0.4,0.4,0.4,1,0,1.4C6.2,6.5,5.9,6.6,5.6,6.6z"></path> <path d="M19.8,20.8c-0.3,0-0.5-0.1-0.7-0.3l-1.4-1.4c-0.4-0.4-0.4-1,0-1.4s1-0.4,1.4,0l1.4,1.4c0.4,0.4,0.4,1,0,1.4C20.3,20.7,20,20.8,19.8,20.8z"></path> <path d="M3,13H1c-0.6,0-1-0.4-1-1s0.4-1,1-1h2c0.6,0,1,0.4,1,1S3.6,13,3,13z"></path> <path d="M23,13h-2c-0.6,0-1-0.4-1-1s0.4-1,1-1h2c0.6,0,1,0.4,1,1S23.6,13,23,13z"></path> <path d="M4.2,20.8c-0.3,0-0.5-0.1-0.7-0.3c-0.4-0.4-0.4-1,0-1.4l1.4-1.4c0.4-0.4,1-0.4,1.4,0s0.4,1,0,1.4l-1.4,1.4C4.7,20.7,4.5,20.8,4.2,20.8z"></path> <path d="M18.4,6.6c-0.3,0-0.5-0.1-0.7-0.3c-0.4-0.4-0.4-1,0-1.4l1.4-1.4c0.4-0.4,1-0.4,1.4,0s0.4,1,0,1.4l-1.4,1.4C18.9,6.5,18.6,6.6,18.4,6.6z"></path></svg> <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class="vt-switch-appearance-moon"><path d="M12.1,22c-0.3,0-0.6,0-0.9,0c-5.5-0.5-9.5-5.4-9-10.9c0.4-4.8,4.2-8.6,9-9c0.4,0,0.8,0.2,1,0.5c0.2,0.3,0.2,0.8-0.1,1.1c-2,2.7-1.4,6.4,1.3,8.4c2.1,1.6,5,1.6,7.1,0c0.3-0.2,0.7-0.3,1.1-0.1c0.3,0.2,0.5,0.6,0.5,1c-0.2,2.7-1.5,5.1-3.6,6.8C16.6,21.2,14.4,22,12.1,22zM9.3,4.4c-2.9,1-5,3.6-5.2,6.8c-0.4,4.4,2.8,8.3,7.2,8.7c2.1,0.2,4.2-0.4,5.8-1.8c1.1-0.9,1.9-2.1,2.4-3.4c-2.5,0.9-5.3,0.5-7.5-1.1C9.2,11.4,8.1,7.7,9.3,4.4z"></path></svg></span></span></button></div> <!----></nav></div></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"><div class="nav-item"><a href="/cosigtdoc/" class="vp-link nav-link">
  Home
  </a></div><div class="nav-item"><a href="/cosigtdoc/contact/contact.html" class="vp-link nav-link">
  Contact
  </a></div><div class="nav-item"><a href="https://github.com/davidebolo1993/cosigt" target="_blank" class="vp-link external nav-link">
  GitHub
  <!----> </a></div> <div class="VPNavBarAppearance appearance" data-v-19d27b0e><button type="button" aria-label="toggle dark mode" class="vt-switch vt-switch-appearance" data-v-19d27b0e><span class="vt-switch-check"><span class="vt-switch-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class="vt-switch-appearance-sun"><path d="M12,18c-3.3,0-6-2.7-6-6s2.7-6,6-6s6,2.7,6,6S15.3,18,12,18zM12,8c-2.2,0-4,1.8-4,4c0,2.2,1.8,4,4,4c2.2,0,4-1.8,4-4C16,9.8,14.2,8,12,8z"></path> <path d="M12,4c-0.6,0-1-0.4-1-1V1c0-0.6,0.4-1,1-1s1,0.4,1,1v2C13,3.6,12.6,4,12,4z"></path> <path d="M12,24c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1s1,0.4,1,1v2C13,23.6,12.6,24,12,24z"></path> <path d="M5.6,6.6c-0.3,0-0.5-0.1-0.7-0.3L3.5,4.9c-0.4-0.4-0.4-1,0-1.4s1-0.4,1.4,0l1.4,1.4c0.4,0.4,0.4,1,0,1.4C6.2,6.5,5.9,6.6,5.6,6.6z"></path> <path d="M19.8,20.8c-0.3,0-0.5-0.1-0.7-0.3l-1.4-1.4c-0.4-0.4-0.4-1,0-1.4s1-0.4,1.4,0l1.4,1.4c0.4,0.4,0.4,1,0,1.4C20.3,20.7,20,20.8,19.8,20.8z"></path> <path d="M3,13H1c-0.6,0-1-0.4-1-1s0.4-1,1-1h2c0.6,0,1,0.4,1,1S3.6,13,3,13z"></path> <path d="M23,13h-2c-0.6,0-1-0.4-1-1s0.4-1,1-1h2c0.6,0,1,0.4,1,1S23.6,13,23,13z"></path> <path d="M4.2,20.8c-0.3,0-0.5-0.1-0.7-0.3c-0.4-0.4-0.4-1,0-1.4l1.4-1.4c0.4-0.4,1-0.4,1.4,0s0.4,1,0,1.4l-1.4,1.4C4.7,20.7,4.5,20.8,4.2,20.8z"></path> <path d="M18.4,6.6c-0.3,0-0.5-0.1-0.7-0.3c-0.4-0.4-0.4-1,0-1.4l1.4-1.4c0.4-0.4,1-0.4,1.4,0s0.4,1,0,1.4l-1.4,1.4C18.9,6.5,18.6,6.6,18.4,6.6z"></path></svg> <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class="vt-switch-appearance-moon"><path d="M12.1,22c-0.3,0-0.6,0-0.9,0c-5.5-0.5-9.5-5.4-9-10.9c0.4-4.8,4.2-8.6,9-9c0.4,0,0.8,0.2,1,0.5c0.2,0.3,0.2,0.8-0.1,1.1c-2,2.7-1.4,6.4,1.3,8.4c2.1,1.6,5,1.6,7.1,0c0.3-0.2,0.7-0.3,1.1-0.1c0.3,0.2,0.5,0.6,0.5,1c-0.2,2.7-1.5,5.1-3.6,6.8C16.6,21.2,14.4,22,12.1,22zM9.3,4.4c-2.9,1-5,3.6-5.2,6.8c-0.4,4.4,2.8,8.3,7.2,8.7c2.1,0.2,4.2-0.4,5.8-1.8c1.1-0.9,1.9-2.1,2.4-3.4c-2.5,0.9-5.3,0.5-7.5-1.1C9.2,11.4,8.1,7.7,9.3,4.4z"></path></svg></span></span></button></div> <!----></nav>  <ul class="sidebar-links"><li><a href="/cosigtdoc/introduction/introduction.html" class="sidebar-link">Introduction</a></li><li><a href="/cosigtdoc/setup/setup.html" class="sidebar-link">Pipeline Setup</a></li><li><a href="/cosigtdoc/workflow/workflow.html" class="sidebar-link">Workflow</a></li><li><a href="/cosigtdoc/usecases/usecases.html" aria-current="page" class="active sidebar-link">Use Cases</a></li><li><a href="/cosigtdoc/contact/contact.html" class="sidebar-link">Contact</a></li></ul> </aside> <main class="page usecasesusecases.html"> <div class="theme-default-content vp-doc content__default"><h1 id="use-cases"><a href="#use-cases" class="header-anchor">#</a> Use Cases</h1> <p>This section provides a comprehensive end-to-end example of how to run the <a href="https://github.com/davidebolo1993/cosigt" target="_blank" rel="noopener noreferrer">cosigt<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> pipeline. The rationale of the pipeline is explained in greater detail in the <a href="/cosigtdoc/workflow/workflow.html">→ workflow section</a>.</p> <h2 id="data-acquisition"><a href="#data-acquisition" class="header-anchor">#</a> Data Acquisition</h2> <p>We create a minimal test dataset to demonstrate cosigt functionality using publicly available data. Let's start by creating a test directory:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token function">mkdir</span> cosigt_test
<span class="token builtin class-name">cd</span> cosigt_test
</code></pre></div><h3 id="required-tools"><a href="#required-tools" class="header-anchor">#</a> Required Tools</h3> <p>Several tools are necessary to complete the following sections. The most straightforward approach is to install them using <a href="https://docs.conda.io/projects/conda/en/latest/index.html" target="_blank" rel="noopener noreferrer">conda<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> or similar package managers, as shown below.</p> <div class="language-bash extra-class"><pre class="language-bash"><code>micromamba create <span class="token punctuation">\</span>
    <span class="token parameter variable">-p</span> use_cases_env <span class="token punctuation">\</span>
    <span class="token parameter variable">-c</span> bioconda <span class="token punctuation">\</span>
    <span class="token parameter variable">-c</span> conda-forge <span class="token punctuation">\</span>
    samtools <span class="token punctuation">\</span>
    odgi <span class="token punctuation">\</span>
    agc <span class="token punctuation">\</span>
    mash <span class="token punctuation">\</span>
    bedtools
micromamba activate <span class="token environment constant">$PWD</span>/use_cases_env
</code></pre></div><h3 id="sequencing-reads"><a href="#sequencing-reads" class="header-anchor">#</a> Sequencing Reads</h3> <p>The cosigt pipeline requires sequencing reads previously aligned to a reference genome in standard <a href="https://samtools.github.io/hts-specs/SAMv1.pdf" target="_blank" rel="noopener noreferrer">bam<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> or <a href="https://samtools.github.io/hts-specs/CRAMv3.pdf" target="_blank" rel="noopener noreferrer">cram<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> format. We can obtain suitable alignment files from the <a href="https://www.internationalgenome.org/data-portal/data-collection/30x-grch38" target="_blank" rel="noopener noreferrer">1000 Genomes Project<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token function">mkdir</span> alignments
<span class="token builtin class-name">cd</span> alignments
<span class="token comment"># Download the complete sample index</span>
<span class="token function">wget</span> https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/additional_698_related/1000G_698_related_high_coverage.sequence.index
<span class="token comment"># Select a couple of samples</span>
<span class="token function">grep</span> <span class="token parameter variable">-E</span> <span class="token string">&quot;HG00438|HG01952&quot;</span> 1000G_698_related_high_coverage.sequence.index <span class="token operator">|</span> <span class="token function">cut</span> <span class="token parameter variable">-f</span> <span class="token number">1</span> <span class="token operator">&gt;</span> 1000G.selected.txt
<span class="token comment"># Add their index files</span>
<span class="token function">sed</span> <span class="token string">'s/$/.crai/'</span> 1000G.selected.txt <span class="token operator">&gt;&gt;</span> 1000G.selected.txt
<span class="token comment"># Download the selected files</span>
<span class="token function">wget</span> <span class="token parameter variable">-i</span> 1000G.selected.txt
<span class="token builtin class-name">cd</span> <span class="token punctuation">..</span>
</code></pre></div><h3 id="reference-genome"><a href="#reference-genome" class="header-anchor">#</a> Reference Genome</h3> <p>A reference genome is required for two purposes: extracting region-specific reads from the original alignments (if in cram format) and serving as one of the assemblies against which samples are compared. In this example, reads are aligned to <a href="https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/" target="_blank" rel="noopener noreferrer">GRCh38<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token function">mkdir</span> reference
<span class="token builtin class-name">cd</span> reference
<span class="token comment"># Download the reference genome</span>
<span class="token function">wget</span> ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa
<span class="token comment"># Download its index</span>
<span class="token function">wget</span> ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa.fai
<span class="token builtin class-name">cd</span> <span class="token punctuation">..</span>
</code></pre></div><h3 id="genome-assemblies"><a href="#genome-assemblies" class="header-anchor">#</a> Genome Assemblies</h3> <p>The cosigt pipeline begins with individual haplotypes to construct pangenome graphs of regions targeted for genotyping. Our workflow expects haplotypes to be grouped by their corresponding reference chromosome. Below are two alternative strategies for generating such chromosome-specific assemblies, which users can choose based on their available data.</p> <h4 id="option-1-starting-from-a-chromosome-specific-graph"><a href="#option-1-starting-from-a-chromosome-specific-graph" class="header-anchor">#</a> Option 1: Starting from a Chromosome-Specific Graph</h4> <p>One approach is to extract haplotypes for specific chromosomes from existing chromosome-specific pangenome graphs. For example, <a href="https://github.com/human-pangenomics/HPP_Year1_Assemblies" target="_blank" rel="noopener noreferrer">Year 1<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> of the <a href="https://humanpangenome.org/" target="_blank" rel="noopener noreferrer">Human Pangenome Reference Consortium (HPRC)<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> provides <a href="https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=pangenomes/freeze/freeze1/pggb/chroms/" target="_blank" rel="noopener noreferrer">such data<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>. We'll limit our analysis to <code>chr4</code> and <code>chr6</code> since these contain our <a href="#regions-of-interest">regions of interest</a>:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token function">mkdir</span> assemblies_a
<span class="token builtin class-name">cd</span> assemblies_a
<span class="token comment"># Download chromosome-specific pangenome graphs</span>
<span class="token function">wget</span> https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/freeze/freeze1/pggb/chroms/chr4.hprc-v1.0-pggb.gfa.gz
<span class="token function">wget</span> https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/freeze/freeze1/pggb/chroms/chr6.hprc-v1.0-pggb.gfa.gz
<span class="token comment"># Download AGC (Assembly Graph Container)</span>
<span class="token function">wget</span> https://zenodo.org/record/5826274/files/HPRC-yr1.agc?download<span class="token operator">=</span><span class="token number">1</span> <span class="token parameter variable">-O</span> HPRC-yr1.agc 
<span class="token comment"># Process each chromosome to subset HG00438, HG01952, plus 8 additional samples</span>
<span class="token keyword">for</span> <span class="token for-or-select variable">chr</span> <span class="token keyword">in</span> chr4 chr6<span class="token punctuation">;</span> <span class="token keyword">do</span>
    <span class="token comment"># Extract all contig IDs</span>
    zgrep <span class="token string">&quot;^P&quot;</span> <span class="token string">&quot;<span class="token variable">$chr</span>.hprc-v1.0-pggb.gfa.gz&quot;</span> <span class="token operator">|</span> <span class="token function">cut</span> <span class="token parameter variable">-f2</span> <span class="token operator">|</span> <span class="token function">sort</span> <span class="token parameter variable">-u</span> <span class="token operator">&gt;</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.all_ids.txt&quot;</span>
    <span class="token comment"># Subset to our samples of interest</span>
    <span class="token function">grep</span> <span class="token parameter variable">-E</span> <span class="token string">&quot;^HG00438#|^HG01952#&quot;</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.all_ids.txt&quot;</span> <span class="token operator">&gt;</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.subset.txt&quot;</span>
    <span class="token comment"># Add 8 additional samples for diversity</span>
    <span class="token function">grep</span> <span class="token parameter variable">-v</span> <span class="token parameter variable">-E</span> <span class="token string">&quot;^HG00438#|^HG01952#&quot;</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.all_ids.txt&quot;</span> <span class="token punctuation">\</span>
        <span class="token operator">|</span> <span class="token function">cut</span> -d<span class="token string">'#'</span> <span class="token parameter variable">-f1</span> <span class="token punctuation">\</span>
        <span class="token operator">|</span> <span class="token function">sort</span> <span class="token parameter variable">-u</span> <span class="token punctuation">\</span>
        <span class="token operator">|</span> <span class="token function">head</span> <span class="token parameter variable">-n</span> <span class="token number">8</span> <span class="token punctuation">\</span>
        <span class="token operator">|</span> <span class="token keyword">while</span> <span class="token builtin class-name">read</span> <span class="token function">id</span><span class="token punctuation">;</span> <span class="token keyword">do</span> <span class="token function">grep</span> <span class="token string">&quot;^<span class="token variable">$id</span>#&quot;</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.all_ids.txt&quot;</span><span class="token punctuation">;</span> <span class="token keyword">done</span> <span class="token operator">&gt;&gt;</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.subset.txt&quot;</span>
    <span class="token function">rm</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.all_ids.txt&quot;</span>
    <span class="token comment"># Extract FASTA sequences</span>
    <span class="token function">cat</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.subset.txt&quot;</span> <span class="token operator">|</span> <span class="token keyword">while</span> <span class="token builtin class-name">read</span> f<span class="token punctuation">;</span> <span class="token keyword">do</span>
        agc getctg HPRC-yr1.agc <span class="token variable">$f</span> <span class="token operator">|</span> bgzip -@ <span class="token number">8</span> <span class="token operator">&gt;&gt;</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.subset.fasta.gz&quot;</span>
    <span class="token keyword">done</span>
    <span class="token function">rm</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.subset.txt&quot;</span>
    <span class="token comment"># Index the FASTA file</span>
    samtools faidx <span class="token string">&quot;<span class="token variable">${chr}</span>.subset.fasta.gz&quot;</span>
<span class="token keyword">done</span>
<span class="token builtin class-name">cd</span> <span class="token punctuation">..</span>
</code></pre></div><div class="custom-block tip"><p class="custom-block-title">TIP</p> <p>This represents just one possible strategy for subsetting assemblies from chromosome-specific pangenome graphs. Alternatively, you can subset the initial .gfa to paths of interest using <code>odgi paths -i GRAPH.GFA -K PATHS.TXT -o SUBGRAPH.OG</code> and then extract the assemblies as FASTA from the subgraph with <code>odgi paths -i SUBGRAPH.OG -f &gt; SUBSET.FASTA</code>.</p></div> <h4 id="option-2-starting-from-individual-haplotypes"><a href="#option-2-starting-from-individual-haplotypes" class="header-anchor">#</a> Option 2: Starting from Individual Haplotypes</h4> <p>An alternative approach is to determine which reference chromosome each sample-specific haplotype most likely belongs to using <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0997-x" target="_blank" rel="noopener noreferrer">Mash distances<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>. This method doesn't require chromosome-specific graphs and can start directly from individual assemblies:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token function">mkdir</span> assemblies_b
<span class="token builtin class-name">cd</span> assemblies_b
<span class="token comment"># Reference to AGC from previous step</span>
<span class="token assign-left variable">agc</span><span class="token operator">=</span><span class="token string">&quot;../assemblies_a/HPRC-yr1.agc&quot;</span>
<span class="token comment"># Extract individual IDs</span>
<span class="token assign-left variable">ids</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token function">cut</span> <span class="token parameter variable">-d</span> <span class="token string">&quot;#&quot;</span> <span class="token parameter variable">-f</span> <span class="token number">1,2</span> <span class="token punctuation">..</span>/assemblies_a/*fai <span class="token operator">|</span> <span class="token function">sed</span> <span class="token string">'s/#/./g'</span> <span class="token operator">|</span> <span class="token function">sort</span> <span class="token operator">|</span> <span class="token function">uniq</span><span class="token variable">)</span></span>
<span class="token comment"># Download corresponding FASTA files</span>
<span class="token keyword">for</span> <span class="token for-or-select variable">id</span> <span class="token keyword">in</span> <span class="token variable">$ids</span><span class="token punctuation">;</span> <span class="token keyword">do</span>
    agc getset <span class="token variable">$agc</span> <span class="token variable">$id</span> <span class="token operator">|</span> bgzip -@ <span class="token number">8</span> <span class="token operator">&gt;</span> <span class="token variable">$id</span>.fasta.gz
    samtools faidx <span class="token variable">$id</span>.fasta.gz
<span class="token keyword">done</span>
<span class="token comment"># Create a sketch of the reference</span>
mash sketch <span class="token parameter variable">-s</span> <span class="token number">100000</span> <span class="token parameter variable">-p</span> <span class="token number">8</span> <span class="token parameter variable">-i</span> <span class="token punctuation">..</span>/reference/GRCh38_full_analysis_set_plus_decoy_hla.fa
<span class="token keyword">for</span> <span class="token for-or-select variable">id</span> <span class="token keyword">in</span> <span class="token variable">$ids</span><span class="token punctuation">;</span> <span class="token keyword">do</span>
    <span class="token comment"># Calculate distances (each contig vs reference chromosomes)</span>
    mash dist <span class="token punctuation">..</span>/reference/GRCh38_full_analysis_set_plus_decoy_hla.fa.msh <span class="token string">&quot;<span class="token variable">$id</span>.fasta.gz&quot;</span> <span class="token parameter variable">-i</span> <span class="token parameter variable">-s</span> <span class="token number">100000</span> <span class="token parameter variable">-p</span> <span class="token number">8</span> <span class="token operator">&gt;</span> <span class="token string">&quot;<span class="token variable">$id</span>.tsv&quot;</span>
    <span class="token comment"># Identify contigs with chr4/chr6 as best matches</span>
    <span class="token function">cut</span> <span class="token parameter variable">-f</span> <span class="token number">2</span> <span class="token string">&quot;<span class="token variable">$id</span>.tsv&quot;</span> <span class="token operator">|</span> <span class="token function">sort</span> <span class="token parameter variable">-u</span> <span class="token operator">|</span> <span class="token keyword">while</span> <span class="token builtin class-name">read</span> f<span class="token punctuation">;</span> <span class="token keyword">do</span>
        <span class="token assign-left variable">best</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token function">grep</span> <span class="token parameter variable">-w</span> <span class="token string">&quot;<span class="token variable">$f</span>&quot;</span> <span class="token string">&quot;<span class="token variable">$id</span>.tsv&quot;</span> <span class="token operator">|</span> <span class="token function">sort</span> <span class="token parameter variable">-k3n</span> <span class="token operator">|</span> <span class="token function">head</span> <span class="token parameter variable">-1</span><span class="token variable">)</span></span>
        <span class="token assign-left variable">chr</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token builtin class-name">echo</span> <span class="token string">&quot;<span class="token variable">$best</span>&quot;</span> <span class="token operator">|</span> <span class="token function">awk</span> <span class="token string">'{print $1}'</span><span class="token variable">)</span></span>
        <span class="token assign-left variable">hap</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token builtin class-name">echo</span> <span class="token string">&quot;<span class="token variable">$best</span>&quot;</span> <span class="token operator">|</span> <span class="token function">awk</span> <span class="token string">'{print $2}'</span><span class="token variable">)</span></span>
        <span class="token keyword">if</span> <span class="token punctuation">[</span><span class="token punctuation">[</span> <span class="token string">&quot;<span class="token variable">$chr</span>&quot;</span> <span class="token operator">==</span> <span class="token string">&quot;chr4&quot;</span> <span class="token operator">||</span> <span class="token string">&quot;<span class="token variable">$chr</span>&quot;</span> <span class="token operator">==</span> <span class="token string">&quot;chr6&quot;</span> <span class="token punctuation">]</span><span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
            samtools faidx <span class="token string">&quot;<span class="token variable">$id</span>.fasta.gz&quot;</span> <span class="token string">&quot;<span class="token variable">$hap</span>&quot;</span> <span class="token operator">|</span> bgzip -@ <span class="token number">8</span> <span class="token operator">&gt;&gt;</span> <span class="token string">&quot;<span class="token variable">${chr}</span>.subset.fasta.gz&quot;</span>
        <span class="token keyword">fi</span>
    <span class="token keyword">done</span>
    <span class="token function">rm</span> <span class="token string">&quot;<span class="token variable">$id</span>.tsv&quot;</span>
<span class="token keyword">done</span>
<span class="token comment"># Index the resulting chromosome-specific files</span>
<span class="token keyword">for</span> <span class="token for-or-select variable">chr</span> <span class="token keyword">in</span> chr4 chr6<span class="token punctuation">;</span> <span class="token keyword">do</span>
    samtools faidx <span class="token string">&quot;<span class="token variable">${chr}</span>.subset.fasta.gz&quot;</span>
<span class="token keyword">done</span>
</code></pre></div><h3 id="regions-of-interest"><a href="#regions-of-interest" class="header-anchor">#</a> Regions of Interest</h3> <p>One final piece of information required for the pipeline is a list of target regions located on the chromosomes for which we have assemblies (chromosomes 6 and 4 in this example). Notable examples include the <a href="https://en.wikipedia.org/wiki/Complement_component_4" target="_blank" rel="noopener noreferrer">C4<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> and <a href="https://en.wikipedia.org/wiki/Glycophorin_A" target="_blank" rel="noopener noreferrer">GYPA<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> loci:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token function">mkdir</span> regions
<span class="token builtin class-name">cd</span> regions
<span class="token comment"># GYPA region</span>
<span class="token builtin class-name">echo</span> <span class="token parameter variable">-e</span> <span class="token string">&quot;chr4<span class="token entity" title="\t">\t</span>144109303<span class="token entity" title="\t">\t</span>144140751<span class="token entity" title="\t">\t</span>GYPA&quot;</span> <span class="token operator">&gt;</span> roi.bed
<span class="token comment"># C4 region</span>
<span class="token builtin class-name">echo</span> <span class="token parameter variable">-e</span> <span class="token string">&quot;chr6<span class="token entity" title="\t">\t</span>31972057<span class="token entity" title="\t">\t</span>32055418<span class="token entity" title="\t">\t</span>C4&quot;</span> <span class="token operator">&gt;&gt;</span> roi.bed
<span class="token builtin class-name">cd</span> <span class="token punctuation">..</span>
</code></pre></div><h3 id="annotations-optional"><a href="#annotations-optional" class="header-anchor">#</a> Annotations (optional)</h3> <p>You can optionally provide gene annotation and protein translation files to the pipeline. If supplied, the pipeline will build gene graphs using <a href="https://github.com/lh3/pangene" target="_blank" rel="noopener noreferrer">pangene<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> allowing gene projection onto the assemblies for the regions of interest.</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token function">mkdir</span> annotations
<span class="token builtin class-name">cd</span> annotations
<span class="token comment"># Download GTF annotation</span>
<span class="token function">wget</span> https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_48/gencode.v48.annotation.gtf.gz
<span class="token comment"># Download and index protein-coding transcript translations</span>
<span class="token function">wget</span> https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_48/gencode.v48.pc_translations.fa.gz
<span class="token function">gzip</span> <span class="token parameter variable">-d</span> gencode.v48.pc_translations.fa.gz
bgzip gencode.v48.pc_translations.fa
samtools faidx gencode.v48.pc_translations.fa.gz
<span class="token builtin class-name">cd</span> <span class="token punctuation">..</span>
</code></pre></div><h3 id="haplotype-exclusion-optional"><a href="#haplotype-exclusion-optional" class="header-anchor">#</a> Haplotype exclusion (optional)</h3> <p>You may optionally provide a list of regions identified by <a href="https://github.com/mobinasri/flagger" target="_blank" rel="noopener noreferrer">flagger<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> in BED format to exclude potentially misassembled haplotype blocks from downstream analyses. The following strategy downloads the Flagger index and corresponding BED files for our assemblies of interest, then generates a unified exclusion file:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token function">mkdir</span> flagger
<span class="token builtin class-name">cd</span> flagger
<span class="token comment"># Download Flagger index</span>
<span class="token function">wget</span> https://raw.githubusercontent.com/human-pangenomics/HPP_Year1_Assemblies/12656b82a42cd4ec6d421abe7fd4ebdeca74b41c/annotation_index/Year1_assemblies_v2_genbank_Flagger.index
<span class="token comment"># Extract assembly names and download corresponding BEDs</span>
<span class="token function">cut</span> <span class="token parameter variable">-f</span> <span class="token number">1</span> <span class="token punctuation">..</span>/assemblies_a/*fai <span class="token operator">|</span> <span class="token function">cut</span> <span class="token parameter variable">-d</span> <span class="token string">&quot;#&quot;</span> <span class="token parameter variable">-f</span> <span class="token number">1</span> <span class="token operator">|</span> <span class="token function">sort</span> <span class="token operator">|</span> <span class="token function">uniq</span> <span class="token operator">|</span> <span class="token keyword">while</span> <span class="token builtin class-name">read</span> f<span class="token punctuation">;</span> <span class="token keyword">do</span>
    <span class="token assign-left variable">link</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token function">grep</span> <span class="token parameter variable">-w</span> $f Year1_assemblies_v2_genbank_Flagger.index <span class="token operator">|</span> <span class="token function">cut</span> <span class="token parameter variable">-f</span> <span class="token number">2</span><span class="token variable">)</span></span>
    <span class="token assign-left variable">clink</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token builtin class-name">echo</span> $link <span class="token operator">|</span> <span class="token function">sed</span> <span class="token string">'s,s3://,https://s3-us-west-2.amazonaws.com/,'</span><span class="token variable">)</span></span>
    <span class="token function">wget</span> <span class="token variable">$clink</span>
<span class="token keyword">done</span>
<span class="token comment"># Merge and sort all exclusion regions</span>
<span class="token function">cut</span> <span class="token parameter variable">-f</span> <span class="token number">1</span>-3 *bed <span class="token operator">|</span> bedtools <span class="token function">sort</span> <span class="token parameter variable">-i</span> - <span class="token operator">&gt;</span> flagger.exclude.bed
<span class="token builtin class-name">cd</span> <span class="token punctuation">..</span>
</code></pre></div><h3 id="clone-cosigt-repository"><a href="#clone-cosigt-repository" class="header-anchor">#</a> Clone cosigt Repository</h3> <p>Clone the cosigt repository from <a href="https://github.com/davidebolo1993/cosigt" target="_blank" rel="noopener noreferrer">GitHub<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>.</p> <div class="language-bash extra-class"><pre class="language-bash"><code>micromamba activate smk7324app132
<span class="token function">git</span> clone https://github.com/davidebolo1993/cosigt
<span class="token builtin class-name">cd</span> cosigt
</code></pre></div><p>For the environment setup, refer to the <a href="/cosigtdoc/setup/setup.html">→ setup section</a>.</p> <h2 id="configure-and-run-the-workflow-master-and-ancient-dna-branches"><a href="#configure-and-run-the-workflow-master-and-ancient-dna-branches" class="header-anchor">#</a> Configure and run the workflow (<code>master</code> and <code>ancient_dna</code> branches)</h2> <p>With all necessary data prepared, we can now configure the pipeline using the <a href="https://github.com/davidebolo1993/cosigt/blob/master/cosigt_smk/workflow/scripts/organize.py" target="_blank" rel="noopener noreferrer">dedicated setup script<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>. The <code>organize.py</code> script of the <code>master</code> and <code>ancient_dna</code> branches of the pipeline accepts the arguments listed below:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token comment"># Confirm you're on the master branch</span>
<span class="token function">git</span> rev-parse --abbrev-ref HEAD
<span class="token comment"># Should return 'master'</span>
<span class="token builtin class-name">cd</span> cosigt_smk
python workflow/scripts/organize.py <span class="token parameter variable">-h</span>
</code></pre></div><div class="language-txt extra-class"><pre class="language-txt"><code>COsine SImilarity-based GenoTyper

optional arguments:
  -h, --help            show this help message and exit

Required I/O arguments:
  -a FASTA, --assemblies FASTA
                        assemblies individuals to -r will be genotyped against. This is a tab-separated file mapping chromosomes in -b to a FASTA with contigs for that chromosome. FASTA can be bgzip-compressed and must be indexed
  -g FASTA, --genome FASTA
                        reference genome. This is the FASTA regions to -b refers to. FASTA can be bgzip-compressed and must be indexed
  -r FOLDER, --reads FOLDER
                        individuals to genotype. This is a folder with individual reads aligned to a reference genome (same to -g) in BAM/CRAM format. Alignment files must be indexed (BAI,CSI/CRAI) and will be searched recursively
  -b BED, --bed BED     regions to genotype. A standard 3-column BED file, but can have a 4th column to label the region and a
                        5th column listing comma-separated alternative contigs for that region
  -o FOLDER, --output FOLDER
                        output folder. This is where results from cosigt pipeline will be stored

Optional arguments:
  --map                 tab-separated file mapping each alignment file (1st column) to an id (2nd column). Guess the name removing file extension otherwise [None]
  --gtf                 Gene annotation on the reference chromosomes in GTF format [None]
  --proteins            Protein-coding transcript translation sequences in FASTA format. FASTA can be bgzip-compressed and must be indexed [None]
  --tmp                 tmp directory. Will be used by tools for which a tmp directory can be specified [/tmp]
  --pansn               PanSN prefix naming for the reference genome [grch38#1#]
  --profile             snakemake profile, if available [None]
  --blacklist           list of contigs to blacklist - these will not be used during genotyping [None]
  --flagger             regions to exclude for each contigs. This is a standard BED file coming from flagger, with contigs names matching those of the assemblies in -a [None]
  --conda               prepare for running using conda instead of singularity [False]
  --threads             run snakemake using that many cores - ignored if using a profile [32]
</code></pre></div><p>For the current use-case, inputs to <code>organize.py</code> are the following:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token comment"># Assemblies map</span>
<span class="token operator">&gt;</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/asm_map.tsv
<span class="token keyword">for</span> <span class="token for-or-select variable">f</span> <span class="token keyword">in</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/assemblies_a/*.fasta.gz<span class="token punctuation">;</span> <span class="token keyword">do</span>
    <span class="token assign-left variable">id</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token function">basename</span> <span class="token string">&quot;<span class="token variable">$f</span>&quot;</span> <span class="token operator">|</span> <span class="token function">cut</span> <span class="token parameter variable">-d</span> <span class="token string">&quot;.&quot;</span> <span class="token parameter variable">-f</span> <span class="token number">1</span><span class="token variable">)</span></span>
    <span class="token builtin class-name">echo</span> <span class="token parameter variable">-e</span> <span class="token string">&quot;<span class="token variable">$id</span><span class="token entity" title="\t">\t</span><span class="token variable">$f</span>&quot;</span> <span class="token operator">&gt;&gt;</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/asm_map.tsv
<span class="token keyword">done</span>
<span class="token assign-left variable">assemblies</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/asm_map.tsv<span class="token variable">)</span></span>
<span class="token comment"># Reference genome</span>
<span class="token assign-left variable">genome</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/reference/GRCh38_full_analysis_set_plus_decoy_hla.fa<span class="token variable">)</span></span>
<span class="token comment"># Read alignments</span>
<span class="token assign-left variable">reads</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/alignments<span class="token variable">)</span></span>
<span class="token comment"># Genotyping regions</span>
<span class="token assign-left variable">bed</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/regions/roi.bed<span class="token variable">)</span></span>
<span class="token comment"># Output directory</span>
<span class="token assign-left variable">output</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token environment constant">$PWD</span>/cosigt_results<span class="token variable">)</span></span>
<span class="token comment"># Sample map (optional)</span>
<span class="token operator">&gt;</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/sample_map.tsv
<span class="token keyword">for</span> <span class="token for-or-select variable">s</span> <span class="token keyword">in</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/alignments/*.cram<span class="token punctuation">;</span> <span class="token keyword">do</span>
    <span class="token assign-left variable">id</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token function">basename</span> <span class="token string">&quot;<span class="token variable">$s</span>&quot;</span> <span class="token operator">|</span> <span class="token function">cut</span> <span class="token parameter variable">-d</span> <span class="token string">&quot;.&quot;</span> <span class="token parameter variable">-f</span> <span class="token number">1</span><span class="token variable">)</span></span>
    <span class="token builtin class-name">echo</span> <span class="token parameter variable">-e</span> <span class="token string">&quot;<span class="token variable">$s</span><span class="token entity" title="\t">\t</span><span class="token variable">$id</span>&quot;</span> <span class="token operator">&gt;&gt;</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/sample_map.tsv
<span class="token keyword">done</span>
<span class="token assign-left variable">map</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/sample_map.tsv<span class="token variable">)</span></span>
<span class="token comment"># GTF (optional)</span>
<span class="token assign-left variable">gtf</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/annotations/gencode.v48.annotation.gtf.gz<span class="token variable">)</span></span>
<span class="token comment"># Proteins FASTA (optional)</span>
<span class="token assign-left variable">proteins</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/annotations/gencode.v48.pc_translations.fa.gz<span class="token variable">)</span></span>
<span class="token comment"># Flagger exclusion (optional)</span>
<span class="token assign-left variable">flagger</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/flagger/flagger.exclude.bed<span class="token variable">)</span></span>
</code></pre></div><p>This generates in the current working directory (<code>cosigt/cosigt_smk</code>):</p> <ul><li>the <code>resources</code> folder</li> <li>the <code>config/config.yaml</code> file</li> <li>the <code>cosigt_smk.sh</code> script</li></ul> <p>In <code>resources</code> input files are re-organized so that the pipeline knows exactly where they are located. They follow the structure below:</p> <div class="language-bash extra-class"><pre class="language-bash"><code>tree resources
</code></pre></div><div class="language-txt extra-class"><pre class="language-txt"><code>resources/
|-- alignments
|   |-- HG00438.cram -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/alignments/HG00438.final.cram
|   |-- HG00438.cram.crai -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/alignments/HG00438.final.cram.crai
|   |-- HG01952.cram -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/alignments/HG01952.final.cram
|   `-- HG01952.cram.crai -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/alignments/HG01952.final.cram.crai
|-- annotations
|   |-- gencode.v48.annotation.gtf.gz -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/annotations/gencode.v48.annotation.gtf.gz
|   |-- gencode.v48.pc_translations.fa.gz -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/annotations/gencode.v48.pc_translations.fa.gz
|   |-- gencode.v48.pc_translations.fa.gz.fai -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/annotations/gencode.v48.pc_translations.fa.gz.fai
|   `-- gencode.v48.pc_translations.fa.gz.gzi -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/annotations/gencode.v48.pc_translations.fa.gz.gzi
|-- assemblies
|   |-- chr4
|   |   |-- chr4.subset.fasta.gz -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/assemblies_a/chr4.subset.fasta.gz
|   |   |-- chr4.subset.fasta.gz.fai -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/assemblies_a/chr4.subset.fasta.gz.fai
|   |   `-- chr4.subset.fasta.gz.gzi -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/assemblies_a/chr4.subset.fasta.gz.gzi
|   `-- chr6
|       |-- chr6.subset.fasta.gz -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/assemblies_a/chr6.subset.fasta.gz
|       |-- chr6.subset.fasta.gz.fai -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/assemblies_a/chr6.subset.fasta.gz.fai
|       `-- chr6.subset.fasta.gz.gzi -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/assemblies_a/chr6.subset.fasta.gz.gzi
|-- blacklist
|   `-- blacklist.txt
|-- flagger
|   `-- flagger_blacklist.bed
|-- reference
|   |-- GRCh38_full_analysis_set_plus_decoy_hla.fa -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/reference/GRCh38_full_analysis_set_plus_decoy_hla.fa
|   `-- GRCh38_full_analysis_set_plus_decoy_hla.fa.fai -&gt; /group/soranzo/davide.bolognini/working/dev/cosigt_test/reference/GRCh38_full_analysis_set_plus_decoy_hla.fa.fai
`-- regions
    |-- all_regions.tsv
    |-- chr4
    |   `-- chr4_144109303_144140751.bed
    `-- chr6
        `-- chr6_31972057_32055418.bed
</code></pre></div><p><code>config/config.yaml</code> defines resources for pipeline components. For example:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token function">grep</span> <span class="token string">&quot;pggb&quot;</span> config/config.yaml
</code></pre></div><div class="language-json extra-class"><pre class="language-json"><code>#used by pggb
pggb<span class="token operator">:</span> <span class="token punctuation">{</span>mem_mb<span class="token operator">:</span> <span class="token number">20000</span><span class="token punctuation">,</span> params<span class="token operator">:</span> -c <span class="token number">2</span><span class="token punctuation">,</span> threads<span class="token operator">:</span> <span class="token number">24</span><span class="token punctuation">,</span> time<span class="token operator">:</span> <span class="token number">40</span><span class="token punctuation">,</span> tmpdir<span class="token operator">:</span> /tmp<span class="token punctuation">}</span>
</code></pre></div><p>This configures <code>pggb</code> to run with 20GB RAM, 24 threads, and 40 minutes time limit using <code>/tmp</code> as a working directory. Resources can be tuned by users based on those available.
<code>cosigt_smk.sh</code> script launches the pipeline. By default, the pipeline will be launched using Singularity for tools. Since we haven’t passed a <code>--profile</code>, the workflow will execute locally.
To run on a SLURM cluster, pass the profile path (see  the <a href="/cosigtdoc/setup/setup.html">→ setup section</a> for details on how to build an example profile for SLURM).</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token comment"># Clean resources</span>
<span class="token function">rm</span> <span class="token parameter variable">-rf</span> resources
<span class="token comment"># Run organize.py</span>
python workflow/scripts/organize.py <span class="token punctuation">\</span>
    <span class="token parameter variable">-a</span> <span class="token variable">$assemblies</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">-g</span> <span class="token variable">$genome</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">-r</span> <span class="token variable">$reads</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">-b</span> <span class="token variable">$bed</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">-o</span> <span class="token variable">$output</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">--map</span> <span class="token variable">$map</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">--gtf</span> <span class="token variable">$gtf</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">--proteins</span> <span class="token variable">$proteins</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">--flagger</span> <span class="token variable">$flagger</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">--profile</span> config/slurm
</code></pre></div><p>The entire workflow can then be run with <code>sh cosigt_smk.sh</code>. Users may want to refine the regions provided to <code>--bed</code> using the <a href="#refine-workflow">refine</a> mode od cosigt. This is not required for with the current example.</p> <h2 id="exploring-results"><a href="#exploring-results" class="header-anchor">#</a> Exploring Results</h2> <p>Following, an overview of the primary outputs produced by the cosigt pipeline</p> <h3 id="genotyping-results"><a href="#genotyping-results" class="header-anchor">#</a> Genotyping Results</h3> <p>Cosigt performs <code>haplotype deconvolution</code>, identifying the combination of haplotypes from the pangenome that best matches the short-read sequencing data for a given region. Details on this step are described in the <a href="/cosigtdoc/workflow/workflow.html">→ workflow section</a>.
After running the pipeline, genotyping outputs are stored under:</p> <div class="language-bash extra-class"><pre class="language-bash"><code>tree cosigt_results/cosigt
</code></pre></div><div class="language-txt extra-class"><pre class="language-txt"><code>cosigt_results/cosigt/
|-- HG00438
|   |-- chr4
|   |   `-- chr4_144109303_144140751
|   |       |-- chr4_144109303_144140751.cosigt_genotype.tsv
|   |       |-- chr4_144109303_144140751.sorted_combos.tsv.gz
|   |       `-- viz
|   |           |-- chr4_144109303_144140751.ava.png
|   |           |-- HG00438#1_to_grch38#1#chr4.png
|   |           `-- HG00438#2_to_grch38#1#chr4.png
|   `-- chr6
|       `-- chr6_31972057_32055418
|           |-- chr6_31972057_32055418.cosigt_genotype.tsv
|           |-- chr6_31972057_32055418.sorted_combos.tsv.gz
|           `-- viz
|               |-- chr6_31972057_32055418.ava.png
|               |-- HG00438#1_to_grch38#1#chr6.png
|               `-- HG00438#2_to_grch38#1#chr6.png
`-- HG01952
    |-- chr4
    |   `-- chr4_144109303_144140751
    |       |-- chr4_144109303_144140751.cosigt_genotype.tsv
    |       |-- chr4_144109303_144140751.sorted_combos.tsv.gz
    |       `-- viz
    |           |-- chr4_144109303_144140751.ava.png
    |           |-- HG01952#1_to_grch38#1#chr4.png
    |           `-- HG01952#2_to_grch38#1#chr4.png
    `-- chr6
        `-- chr6_31972057_32055418
            |-- chr6_31972057_32055418.cosigt_genotype.tsv
            |-- chr6_31972057_32055418.sorted_combos.tsv.gz
            `-- viz
                |-- chr6_31972057_32055418.ava.png
                |-- HG01952#1_to_grch38#1#chr6.png
                `-- HG01952#2_to_grch38#1#chr6.png

</code></pre></div><p>Each region directory contains:</p> <ul><li><code>*.cosigt_genotype.tsv</code>: best-matching haplotype pair (<code>haplotype.1</code>, <code>haplotype.2</code>) selected to maximize the <a href="https://en.wikipedia.org/wiki/Cosine_similarity" target="_blank" rel="noopener noreferrer">cosine similarity<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> to the sample.</li> <li><code>*.sorted_combos.tsv.gz</code>: cosine similarity scores for all tested haplotype combinations.</li> <li><code>viz/</code>: visualizations of alignments between predicted haplotypes and the reference.
One example for <code>HG00438</code> at the C4 locus (<code>chr6_31972057_32055418</code>):</li></ul> <p><a href="./ava.png"><img src="/cosigtdoc/assets/img/ava.3cec7bf8.png" width="650" style="display:block;margin:0 auto;"></a></p> <p>In this case, both predicted haplotypes (<code>HG00438#1</code>, <code>HG00438#2</code>) are structurally similar to each other but diverge from the reference (<code>grch38#chr6</code>), due to a duplication absent in the predicted haplotypes.</p> <h3 id="structural-clustering"><a href="#structural-clustering" class="header-anchor">#</a> Structural Clustering</h3> <p>To summarize major structural variation patterns, cosigt performs clustering of haplotypes based on graph similarity (see the <a href="/cosigtdoc/workflow/workflow.html">→ workflow section</a>). Outputs are stored in:</p> <div class="language-bash extra-class"><pre class="language-bash"><code>tree cosigt_results/cluster
</code></pre></div><div class="language-txt extra-class"><pre class="language-txt"><code>cosigt_results/cluster/
|-- chr4
|   `-- chr4_144109303_144140751
|       |-- chr4_144109303_144140751.clusters.hapdist.norm.tsv
|       |-- chr4_144109303_144140751.clusters.hapdist.tsv
|       |-- chr4_144109303_144140751.clusters.json
|       |-- chr4_144109303_144140751.clusters.medoids.tsv
|       |-- chr4_144109303_144140751.clusters.metrics.tsv
|       `-- chr4_144109303_144140751.clusters.tsv
`-- chr6
    `-- chr6_31972057_32055418
        |-- chr6_31972057_32055418.clusters.hapdist.norm.tsv
        |-- chr6_31972057_32055418.clusters.hapdist.tsv
        |-- chr6_31972057_32055418.clusters.json
        |-- chr6_31972057_32055418.clusters.medoids.tsv
        |-- chr6_31972057_32055418.clusters.metrics.tsv
        `-- chr6_31972057_32055418.clusters.tsv
</code></pre></div><p>File descriptions:</p> <ul><li><code>*.clusters.tsv</code>/<code>*.clusters.json</code>: assigned <code>haplotype.group</code> per <code>haplotype.name</code> (TSV and JSON formats).</li> <li><code>*.clusters.hapdist.tsv</code>: pairwise haplotype distances between groups.</li> <li><code>*.clusters.hapdist.norm.tsv</code>: normalized version of the above (scaled to max pairwise distance).</li> <li><code>*.clusters.medoids.tsv</code>: representative haplotype (medoid) for each group.</li> <li><code>*.clusters.metrics.tsv</code>: clustering parameters (e.g., eps) used.</li></ul> <p>Visualizations of clusters can be found in <code>cosigt_test/odgi/viz/{chromosome}/{region}</code>. An example for the C4 locus:</p> <p><a href="./viz.png"><img src="/cosigtdoc/assets/img/viz.2ef168a3.png" width="650" style="display:block;margin:0 auto;"></a></p> <p>Nodes are oriented relative to the reference (<code>grch38#1#chr6</code>) and colored by path coverage (based on the <a href="https://loading.io/color/feature/Spectral-11/" target="_blank" rel="noopener noreferrer">Spectral-11<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> palette). Haplotypes are grouped and ordered by cluster, facilitating visual comparison of structural similarity.</p> <h3 id="genes-annotation"><a href="#genes-annotation" class="header-anchor">#</a> Genes annotation</h3> <p>If gene annotations and protein sequences were provided, cosigt uses pangene to build gene graphs and project annotations onto the assemblies. An example visualization for the C4 locus:</p> <p><a href="./pangene.png"><img src="/cosigtdoc/assets/img/pangene.75fa9b58.png" width="650" style="display:block;margin:0 auto;"></a></p> <h3 id="interactive-plot-of-genotyping-results"><a href="#interactive-plot-of-genotyping-results" class="header-anchor">#</a> Interactive plot of genotyping results</h3> <p>The cosigt pipeline includes an interactive <a href="https://github.com/davidebolo1993/cosigt/blob/master/cosigt_smk/workflow/scripts/app.r" target="_blank" rel="noopener noreferrer">visualization tool<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> that enables dynamic exploration of genotyping outcomes across samples and regions. You can launch the app as follows:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token builtin class-name">cd</span> workflow/scripts
<span class="token function">bash</span> runapp.sh <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/cosigt_test
<span class="token comment">#Listening on http://0.0.0.0:3838</span>
</code></pre></div><p>Then, open your web browser and navigate to <code>http://0.0.0.0:3838</code> to view the results. For example:</p> <p><a href="./app.png"><img src="/cosigtdoc/assets/img/app.dae996c4.png" width="650" style="display:block;margin:0 auto;"></a></p> <p>The app primarily visualizes short-read coverage (<code>Sample Vector Coverage</code>, blue dots, left y-axis) versus haplotype coverage (<code>Pangenome Vector Coverage</code>, orange dots, right y-axis) across graph nodes (<code>Cumulative Haplotype Length</code>, x-axis). Each dot represents a graph node, and the <code>Cumulative Haplotype Length</code> accounts for the length of each node.
For each region of interest (selectable via the <code>Region</code> menu) and sample (via the <code>Sample</code> menu), you can display up to two short-read vs haplotype coverage comparisons: the first in the upper panel and the second in the lower panel. This allows you to compare the cosigt genotyping prediction against the expected haplotype coverage, helping to identify matches or coverage anomalies.</p> <h2 id="configure-and-run-the-workflow-custom-allels-branch"><a href="#configure-and-run-the-workflow-custom-allels-branch" class="header-anchor">#</a> Configure and run the workflow (<code>custom_allels</code> branch)</h2> <p>As an alternative to running from raw assemblies, cosigt can also be launched directly from a predefined set of alleles — for example, those generated in a previous run:</p> <div class="language-bash extra-class"><pre class="language-bash"><code>tree cosigt_results/bedtools/getfasta
</code></pre></div><div class="language-txt extra-class"><pre class="language-txt"><code>cosigt_results/bedtools/getfasta/
|-- chr4
|   `-- chr4_144109303_144140751
|       |-- chr4_144109303_144140751.fasta.gz
|       |-- chr4_144109303_144140751.fasta.gz.fai
|       `-- chr4_144109303_144140751.fasta.gz.gzi
`-- chr6
    `-- chr6_31972057_32055418
        |-- chr6_31972057_32055418.fasta.gz
        |-- chr6_31972057_32055418.fasta.gz.fai
        `-- chr6_31972057_32055418.fasta.gz.gzi
</code></pre></div><p>To use this mode, switch to the <code>custom_alleles</code> branch:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token builtin class-name">cd</span> <span class="token punctuation">..</span>/
<span class="token function">git</span> checkout custom_alleles
<span class="token builtin class-name">cd</span> cosigt_smk
python workflow/scripts/organize.py <span class="token parameter variable">-h</span>
</code></pre></div><p>The script accepts a similar set of arguments to the <code>master</code> and <code>ancient_dna</code> branches. The key difference lies in the <code>-a</code> argument, which now expects a TSV mapping each region to its corresponding allele FASTA file (can be bgzipped). Example:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token operator">&gt;</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/all_map.tsv
<span class="token keyword">for</span> <span class="token for-or-select variable">f</span> <span class="token keyword">in</span> cosigt_results/bedtools/getfasta/*/*/*gz<span class="token punctuation">;</span> <span class="token keyword">do</span>
    <span class="token assign-left variable">id</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token function">basename</span> <span class="token punctuation">$(</span>dirname $f<span class="token punctuation">)</span><span class="token variable">)</span></span>
    <span class="token builtin class-name">echo</span> <span class="token parameter variable">-e</span> <span class="token string">&quot;<span class="token variable">$id</span><span class="token entity" title="\t">\t</span><span class="token variable">$f</span>&quot;</span> <span class="token operator">&gt;&gt;</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/all_map.tsv
<span class="token keyword">done</span>
</code></pre></div><p>Then run:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token comment"># Clean previous resources</span>
<span class="token function">rm</span> <span class="token parameter variable">-rf</span> resources
<span class="token comment"># Alleles</span>
<span class="token assign-left variable">alleles</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token punctuation">..</span>/<span class="token punctuation">..</span>/all_map.tsv<span class="token variable">)</span></span>
<span class="token comment"># New output directory</span>
<span class="token assign-left variable">output</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span>readlink <span class="token parameter variable">-f</span> <span class="token environment constant">$PWD</span>/cosigt_results_alleles<span class="token variable">)</span></span>
<span class="token comment"># Launch organize.py</span>
python workflow/scripts/organize.py <span class="token punctuation">\</span>
    <span class="token parameter variable">-a</span> <span class="token variable">$alleles</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">-g</span> <span class="token variable">$genome</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">-r</span> <span class="token variable">$reads</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">-b</span> <span class="token variable">$bed</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">-o</span> <span class="token variable">$output</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">--map</span> <span class="token variable">$map</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">--gtf</span> <span class="token variable">$gtf</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">--proteins</span> <span class="token variable">$proteins</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">--flagger</span> <span class="token variable">$flagger</span> <span class="token punctuation">\</span>
    <span class="token parameter variable">--profile</span> config/slurm
</code></pre></div><p>And finally:</p> <div class="language-bash extra-class"><pre class="language-bash"><code><span class="token function">sh</span> cosigt_smk.sh
</code></pre></div><p>The output structure and interpretation remain identical to those described in the <a href="#exploring-results">previuos sections</a>.</p> <h2 id="refine-workflow"><a href="#refine-workflow" class="header-anchor">#</a> Refine Workflow</h2> <p>In addition to the main <strong>cosigt</strong> workflow, our Snakemake pipeline also supports a <strong>refine</strong> mode (available in both the <code>master</code> and <code>ancient_dna</code> branches).<br>
This mode attempts to dynamically extend the input regions at both flanks to retain as many haplotypes as possible during the <code>impg</code> step, up to a configurable extension threshold (default: ±60 kb).
To run the refine workflow, simply replace <code>&quot;cosigt&quot;</code> with <code>&quot;refine&quot;</code> in the <code>cosigt_smk.sh</code> script.
The output is a BED file containing the refined regions in <code>cosigt_results/refine/regions_refined.bed</code>.
This refined BED file can then be used as input for a new cosigt run by replacing the original regions BED file.<br>
If you reuse the same output directory (<code>--output</code>), the pipeline will skip the initial query-to-target alignment step, recycling those generated in the <code>refine</code> mode.</p></div> <footer class="page-edit"><!----> <!----></footer> <div class="page-nav"><div class="inner"><a href="/cosigtdoc/workflow/workflow.html" class="vp-link prev"><span>←</span>
  Workflow
  <span><br> <span class="prev-link">/workflow/workflow.html</span></span></a> <a href="/cosigtdoc/contact/contact.html" class="vp-link next">
  Contact
  <span>
        →
        <br> <span class="next-link">/contact/contact.html</span></span></a></div></div> </main> <div class="sticker vuepress-toc"><div class="on-this-page">ON THIS PAGE</div> <div class="vuepress-toc-item vuepress-toc-h2 active"><a href="#data-acquisition" title="Data Acquisition">Data Acquisition</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#required-tools" title="Required Tools">Required Tools</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#sequencing-reads" title="Sequencing Reads">Sequencing Reads</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#reference-genome" title="Reference Genome">Reference Genome</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#genome-assemblies" title="Genome Assemblies">Genome Assemblies</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#regions-of-interest" title="Regions of Interest">Regions of Interest</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#annotations-optional" title="Annotations (optional)">Annotations (optional)</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#haplotype-exclusion-optional" title="Haplotype exclusion (optional)">Haplotype exclusion (optional)</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#clone-cosigt-repository" title="Clone cosigt Repository">Clone cosigt Repository</a></div><div class="vuepress-toc-item vuepress-toc-h2"><a href="#configure-and-run-the-workflow-master-and-ancient-dna-branches" title="Configure and run the workflow (master and ancient_dna branches)">Configure and run the workflow (master and ancient_dna branches)</a></div><div class="vuepress-toc-item vuepress-toc-h2"><a href="#exploring-results" title="Exploring Results">Exploring Results</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#genotyping-results" title="Genotyping Results">Genotyping Results</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#structural-clustering" title="Structural Clustering">Structural Clustering</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#genes-annotation" title="Genes annotation">Genes annotation</a></div><div class="vuepress-toc-item vuepress-toc-h3"><a href="#interactive-plot-of-genotyping-results" title="Interactive plot of genotyping results">Interactive plot of genotyping results</a></div><div class="vuepress-toc-item vuepress-toc-h2"><a href="#configure-and-run-the-workflow-custom-allels-branch" title="Configure and run the workflow (custom_allels branch)">Configure and run the workflow (custom_allels branch)</a></div><div class="vuepress-toc-item vuepress-toc-h2"><a href="#refine-workflow" title="Refine Workflow">Refine Workflow</a></div></div></div><div class="global-ui"></div></div>
    <script src="/cosigtdoc/assets/js/app.ea2c52ed.js" defer></script><script src="/cosigtdoc/assets/js/3.d0293945.js" defer></script><script src="/cosigtdoc/assets/js/2.bc6597f8.js" defer></script><script src="/cosigtdoc/assets/js/1.50924563.js" defer></script><script src="/cosigtdoc/assets/js/14.3e4a7b1f.js" defer></script><script src="/cosigtdoc/assets/js/29.296314f5.js" defer></script>
  </body>
</html>
